### PR TITLE
Array filtering support (Part #4): Postgres array filter impl.

### PR DESCRIPTION
**Pull request by suresh-prakash**
So far, we do not have support for filtering by LHS arrays.
The PRs in this series aim to support LHS arrays in filters.

An LHS array can contain either primitives or documents.
In the case of a primitive LHS array, we can apply a relational operator on each element.
In the case of a document LHS array, we can apply a generic filter (including relational expression or logical expression or even another array filter expression)

Sample data: 
```json
[
  {
    "key": 3,
    "planets": [
      {
        "name": "Planet 1",
        "elements": [
          "Oxygen",
          "Water",
          "Nitrogen"
        ]
      },
      {
        "name": "Planet 2",
        "elements": [
          "Oxygen",
          "Helium",
          "Water"
        ]
      }
    ]
  },
  {
    "key": 1,
    "planets": [
      {
        "name": "Mercury"
      },
      {
        "name": "Venus"
      },
      {
        "name": "Earth",
        "elements": [
          "Oxygen",
          "Nitrogen"
        ]
      },
      {
        "name": "Mars",
        "elements": [
          "Iron"
        ]
      },
      {
        "name": "Jupiter",
        "elements": []
      },
      {
        "name": "Saturn",
        "elements": []
      },
      {
        "name": "Uranus",
        "elements": [
          "Hydrogen",
          "Helium",
          "Methane"
        ]
      },
      {
        "name": "Neptune",
        "elements": [
          "Hydrogen",
          "Helium",
          "Methane",
          "Ammonia"
        ]
      }
    ]
  },
  {
    "key": 2,
    "planets": [
      {
        "name": "Mercury",
        "elements": [
          "Silicate",
          "Aluminum"
        ]
      },
      {
        "name": "Venus",
        "elements": [
          "Carbon Dioxide",
          "Sulfuric Acid"
        ]
      },
      {
        "name": "Earth",
        "elements": [
          "Oxygen",
          "Nitrogen",
          "Water"
        ]
      },
      {
        "name": "Mars",
        "elements": [
          "Iron",
          "Silicate"
        ]
      },
      {
        "name": "Jupiter",
        "elements": [
          "Hydrogen",
          "Helium"
        ]
      },
      {
        "name": "Saturn",
        "elements": [
          "Hydrogen",
          "Helium",
          "Methane"
        ]
      },
      {
        "name": "Uranus",
        "elements": [
          "Hydrogen",
          "Helium",
          "Methane",
          "Ammonia"
        ]
      },
      {
        "name": "Neptune",
        "elements": [
          "Hydrogen",
          "Helium",
          "Methane",
          "Ammonia",
          "Methane Hydrate"
        ]
      }
    ]
  }
]
```

Input query: `Get a list of solar systems with at least one planet not having either oxygen or water.`
```sql
SELECT * 
FROM solar_systems 
WHERE ANY(planets) [NOT ((ANY(elements) = "Oxygen") OR (ANY(elements) = "Water"))]
```

The filter API would look like
```java
FilterTypeExpression filter = DocumentArrayFilterExpression.builder()
    .operator(ANY)
    .arraySource(IdentifierExpression.of("planets"))
    .filter(
       LogicalExpression.builder()
           .operator(NOT)
           .operand(
                 LogicalExpression.builder()
                       .operator(OR)
                       .operand(
                             ArrayRelationalFilterExpression.builder()
                                  .operator(ANY)
                                  .filter(
                                       RelationalExpression.of(
                                            IdentifierExpression.of("elements"),
                                            EQ,
                                            ConstantExpression.of("Oxygen"))))
                        .operand(
                             ArrayRelationalFilterExpression.builder()
                                  .operator(ANY)
                                  .filter(
                                       RelationalExpression.of(
                                            IdentifierExpression.of("elements"),
                                            EQ,
                                            ConstantExpression.of("Water"))))))
    .build();
```

Sample MongoDB match stage query to be generated
```json
{
  "$expr":
  {
    "$anyElementTrue":
    {
      "$map":
      {
        "input":
        {
          "$ifNull": [
            "$planets",
            []
          ]
        },
        "as": "planet",
        "in":
        {
          "$not":
          {
            "$or": [
              {
                "$anyElementTrue":
                {
                  "$map":
                  {
                    "input":
                    {
                      "$ifNull": [
                        "$$planet.elements",
                        []
                      ]
                    },
                    "as": "element",
                    "in":
                    {
                      "$eq": ["$$element","Oxygen"]
                    }
                  }
                }
              },
              {
                "$anyElementTrue":
                {
                  "$map":
                  {
                    "input":
                    {
                      "$ifNull": [
                        "$$planet.elements",
                        []
                      ]
                    },
                    "as": "element",
                    "in":
                    {
                      "$eq": ["$$element","Water"]
                    }
                  }
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

Sample Postgres Query:
```sql
SELECT *  
FROM solar_systems  
WHERE 
EXISTS 
(SELECT 1 
 FROM  jsonb_array_elements(COALESCE(str->'planets', '[]'::jsonb)) AS planet 
 WHERE 
 NOT 
 (
 EXISTS 
 (SELECT 1 
  FROM jsonb_array_elements(COALESCE(planet->'elements', '[]'::jsonb)) AS elements 
  WHERE TRIM('"' FROM elements::text) = 'Oxygen'
 )
 OR 
 EXISTS 
 (SELECT 1 
  FROM jsonb_array_elements(COALESCE(planet->'elements', '[]'::jsonb)) AS elements 
  WHERE TRIM('"' FROM elements::text) = 'Water'
 )
 )
);
```

**Notes:**

1. Only `ANY` operator is supported for now
2. A future PR would introduce a `NOT` logical operator
3. `ALL` can be achieved by using `NOT` twice to the `ANY` filter (before and after). E.g.: `Fetch solar systems with all planets containing both oxygen and water` is equivalent to `Fetch solar systems with no planet without both oxygen and water`
4. Similarly `NONE` can be achieved by using `NOT` once before `ANY`, because `NONE = NOT ANY`
5. The examples given in this description are for illustrative purposes only. For practical applications, the queries would be simpler (E.g.: Use `IN` instead of `OR` and `=`, static import `not` and `or` logical operators, etc.)

**This PR contains**
* Postgres array filter implementation

**Previous PRs**
#173 
#188 
#189

### CODE REVIEW

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresIdentifierAccessingExpressionVisitor.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  public PostgresIdentifierAccessingExpressionVisitor(
+      final PostgresSelectTypeExpressionVisitor baseVisitor, final String baseField) {
+    super(baseVisitor);
+    this.baseField = baseField;
+  }
+
+  @Override
+  public PostgresQueryParser getPostgresQueryParser() {
+    return baseVisitor.getPostgresQueryParser();
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresIdentifierAccessingExpressionVisitor.java:21]: Curious: In the `PostgresIdentifierReplacingExpressionVisitor` class, we initially verify the existence of the `postgresQueryParser`. Afterward, we reference the `baseVisitor` as follows:

```java
postgresQueryParser != null ? postgresQueryParser : baseVisitor.getPostgresQueryParser()
```

Why aren't we doing the same here and directly accessing it from the `baseVisitor`?

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresRelationalIdentifierAccessingExpressionVisitor.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      final String baseField,
+      final SelectTypeExpression rhs) {
+    super(baseVisitor);
+    this.baseField = baseField;
+    this.rhs = rhs;
+  }
+
+  @Override
+  public PostgresQueryParser getPostgresQueryParser() {
+    return baseVisitor.getPostgresQueryParser();
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresRelationalIdentifierAccessingExpressionVisitor.java:26]: Same question as above: https://github.com/hypertrace/document-store/pull/191#discussion_r1452928412

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java [hunk 3]
```diff
@@ -81,16 +96,108 @@ public String visit(final KeyExpression expression) {
...
+      FROM jsonb_array_elements(COALESCE(planets->'elements', '[]'::jsonb)) AS elements
+      WHERE TRIM('"' FROM elements::text) = 'Oxygen'
+     )
+     */
+    switch (expression.getOperator()) {
+      case ANY:
+        // Convert 'elements' to planets->'elements' where planets could be an alias for an upper
+        // level array filter
+        // Also, for the first time (if this was not under any nesting), use the field identifier
+        // visitor to make it document->'elements'
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java:114]: is this `document->'elements'` correct? should it be `document->'planets'?

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java [hunk 4]
```diff
@@ -81,16 +96,108 @@ public String visit(final KeyExpression expression) {
...
+                : wrappingVisitorProvider.getForNonRelational(identifierVisitor);
+        final String parsedLhs = expression.getArraySource().accept(arrayPathVisitor);
+
+        // Extract the field name
+        final String identifierName =
+            expression
+                .getArraySource()
+                .accept(new PostgresIdentifierExpressionVisitor(postgresQueryParser));
+
+        // If the field name is 'elements.inner', just pick the last part as the alias ('inner')
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java:182]: Could this be an issue if there are two array expressions, both having the same last part, such as `inner` in `planets.inner` and `elements.inner`?

If so, creating an alias with the full path, like `elements_inner`, might be a helpful solution.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java [hunk 5]
```diff
@@ -81,16 +96,108 @@ public String visit(final KeyExpression expression) {
...
-    throw new UnsupportedOperationException();
+    /*
+    EXISTS
+     (SELECT 1
+      FROM jsonb_array_elements(COALESCE(planets->'elements', '[]'::jsonb)) AS elements
+      WHERE TRIM('"' FROM elements::text) = 'Oxygen'
+     )
+     */
+    switch (expression.getOperator()) {
+      case ANY:
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java:110]: If there are more operators, we should move the code for each operator to its dedicated function for readability? For instance, moving code of any operator to `getFilterExpressionForAnyOperator()` here.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresIdentifierAccessingExpressionVisitor.java (hunk 1 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresIdentifierAccessingExpressionVisitor.java:21]: Actually, that is also not required as we are not passing in any `baseVisitor` in the constructor. Updated there to get rid of the unnecessary conditional statement.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresRelationalIdentifierAccessingExpressionVisitor.java (hunk 2 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresRelationalIdentifierAccessingExpressionVisitor.java:26]: Actually, that is also not required as we are not passing in any baseVisitor in the constructor. Updated there to get rid of the unnecessary conditional statement.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java (hunk 3 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java:108]: `document->'elements'` is correct. I've updated the code comments to be more clear.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java (hunk 4 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java:176]: Actually, the alias is used within a smaller scope, so it shouldn't be an issue. But, yeah, it still makes sense to update it for clarity. Will do.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java (hunk 5 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/vistors/PostgresFilterTypeExpressionVisitor.java:104]: Sure

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/utils/PostgresUtils.java [hunk 6]
```diff
@@ -547,6 +547,11 @@ public static String wrapAliasWithDoubleQuotes(String fieldName) {
     return "\"" + fieldName + "\"";
   }
 
+  public static String getLastPath(final String fieldName) {
+    final String[] split = fieldName.split(DOC_PATH_SEPARATOR);
+    return split[split.length - 1];
+  }
+
```
Comment (suresh-prakash) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/utils/PostgresUtils.java:554]:
```suggestion

```

**Pull request merged**